QccPack-0.61-1/src/libQccPackWAV/utils/man1/srencode.1
.TH SRENCODE 1 "QCCPACK" ""
.SH NAME
srencode, srdecode \-
encode/decode an image using the SR algorithm
.SH SYNOPSIS
.sp
.B srencode
.RB "[\|" \-w
.IR  wavelet "\|]"
.RB "[\|" \-b
.IR  boundary "\|]"
.RB "[\|" \-pw
.IR  pcpfile "\|]"
.RB "[\|" \-nl
.IR  levels "\|]"
.RB "[\|" \-q
.IR  stepsize "\|]"
.RB "[\|" \-d
.IR  dead\_zone "\|]"
.RB "[\|" \-vo "\|]"
.I imgfile
.I bitstream
.LP
.B srdecode
.RB "[\|" \-w
.IR  wavelet "\|]"
.RB "[\|" \-b
.IR  boundary "\|]"
.RB "[\|" \-pw
.IR  pcpfile "\|]"
.I bitstream
.I imgfile
.SH OPTIONS
.TP
.BI \-w " wavelet"
String. 
Name of the wavelet file
.RB ( FBK
or
.B LFT
format); default = "CohenDaubechiesFeauveau.9-7.lft"
.TP
.BI \-b " boundary"
String. Type of boundary extension; default = "symmetric".
.TP
.BI \-pw " pcpfile"
String. Name of file containing perceptual weights.
.TP 
.BI \-nl " levels"
Integer. Number of levels of dyadic decomposition; default value is 5 levels.
.TP 
.BI \-q " stepsize"
Float. Stepsize of the uniform scalar quantizer. Default value is 47.1.
.TP
.BI \-d " dead\_zone"
Float. Size of the dead zone to use with the scalar quantizer.  Default is
to not use a dead zone.
.TP
.B \-vo
Print only the value of the rate to stdout (terse output);
default is verbose output.
.SH DESCRIPTION
.LP
.B srencode
codes an image,
.IR imgfile ,
using the stack-run (SR) algorithm by Tsai 
.IR "et al" .
The SR algorithm involves a 2D DWT followed by 
uniform scalar quantization of the wavelet coefficients.
Runs of the zero index are identified and are run-length encoded.
The SR algorithm then uses a 4-symbol alphabet to code the binary digits
of the runs and the quantizer indices of the nonzero coefficients
(the "stacks").  The stream of symbols from this 4-symbol alphabet
is then coded with arithmetic entropy coding.
.LP
The DWT is performed as described for
.BR imgdwt (1).
Option
.B \-w
specifies the wavelet, 
option
.B \-b
specifies the boundary extension, and option
.B \-pw
indicates whether perceptual weighting is to be used.
See 
.BR imgdwt (1)
for more details.
.LP
The rate-distortion performance of the SR algorithm is determined by the
stepsize of the uniform scalar quantizer, specified
with the
.B \-q
option.
Additionally, an optional dead zone may be applied to the uniform scalar
quantizer.  That is, the size of the quantizer bin centered at zero may
be enlarged so that more wavelet coefficients are quantized to zero,
consequently increasing the likelihood of longer runs (see
.BR squniform (1)
for more details).
.LP
.B srencode
normally prints to stdout a small amount of information
concerning the encoding being performed, most notably the bit rate,
expressed in bits per pixel,
achieved by the program.  If the 
.B \-vo
(value-only) option is given, this printed output is suppressed except for
the bit rate.
.LP
Note: the original presentation of SR given by Tsai 
.IR "et al" .
does not include perceptual weighting as allowed by
.B srencode
(the 
.B \-pw
option).  Refer to
.BR imgdwt (1)
and
.BR QccWAVPerceptualWeights (3)
for details on the implementation of perceptual weighting.
.LP
Note:
.BR srencode
currently achieves rate-distortion performance slightly below that reported
by Tsai 
.IR "et al" .
For instance, when coding the "Barbara" 
image at 0.5 bits/pixel, 
.B srencode 
achieves a PSNR about 0.5dB below the reported figure.
The reasons for this performance gap are currently being investigated.
.LP
.B srdecode
performs the corresponding decoding using the SR algorithm, decoding
the
.I bitstream
made by
.B srencode
and producing
a reconstructed image,
.IR imgfile .
For sane results, 
.IR wavelet ,
.IR boundary ,
and
.I pcpfile
should be the same as they were for
.BR srencode .
.SH "SEE ALSO"
.BR QccWAVsrEncode (3),
.BR QccWAVsrDecode (3),
.BR imgdwt (1),
.BR squniform (1),
.BR QccWAVPerceptualWeights (3),
.BR QccWAVSubbandPyramid (3),
.BR QccPackWAV (3),
.BR QccPackIMG (3),
.BR QccPack (3)

M.-J. Tsai, J. D. Villasenor, and F. Chen,
"Stack-Run Image Coding,"
.IR "IEEE Transactions on Circuits and Systems for Video Technology" ,
vol. 6, no. 5, pp. 519-521, October 1996.

M.-J. Tsai, J. D. Villasenor, and F. Chen,
"Stack-Run Coding for Low Bit Rate Image Communication,"
.IR "Proceedings IEEE International Conference on Image Processing" ,
pp. 681-684, September 1996.

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackWAV/utils/srencode.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#include "srencode.h"

#define USG_STRING "[-w %s:wavelet] [-b %s:boundary] [-pw %: %s:pcpfile] [-nl %d:num_levels] [-q %f:stepsize] [-d %f:deadzone] [-vo %:] %s:imgfile %s:bitstream"

QccWAVWavelet Wavelet;
QccString WaveletFilename = QCCWAVWAVELET_DEFAULT_WAVELET;
QccString Boundary = "symmetric";

QccWAVPerceptualWeights PerceptualWeights;
int UsePerceptualWeights = 0;

int NumLevels = 5;
int NumSubbands;

float StepSize = 47.1;
float DeadZone = -1;
QccSQScalarQuantizer Quantizer;

QccIMGImage InputImage;
int ImageNumRows, ImageNumCols;

QccBitBuffer OutputBuffer;

int ValueOnly = 0;

double Rate;

int main(int argc, char *argv[])
{

  QccInit(argc, argv);
  
  QccWAVWaveletInitialize(&Wavelet);
  QccWAVPerceptualWeightsInitialize(&PerceptualWeights);
  QccIMGImageInitialize(&InputImage);
  QccBitBufferInitialize(&OutputBuffer);
  QccSQScalarQuantizerInitialize(&Quantizer);

  if (QccParseParameters(argc, argv,
                         USG_STRING,
                         WaveletFilename,
                         Boundary,
                         &UsePerceptualWeights,
                         PerceptualWeights.filename,
                         &NumLevels,
                         &StepSize,
                         &DeadZone,
                         &ValueOnly,
                         InputImage.filename,
                         OutputBuffer.filename))
    QccErrorExit();
  
  if (NumLevels < 0)
    {
      QccErrorAddMessage("%s: Number of levels of decomposition must be nonnegative",
                         argv[0]);
      QccErrorExit();
    }
  NumSubbands = QccWAVSubbandPyramidNumLevelsToNumSubbands(NumLevels);
  
  if (QccWAVWaveletCreate(&Wavelet, WaveletFilename, Boundary))
    {
      QccErrorAddMessage("%s: Error calling QccWAVWaveletCreate()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (UsePerceptualWeights)
    {
      if (QccWAVPerceptualWeightsRead(&PerceptualWeights))
        {
          QccErrorAddMessage("%s: Error calling QccWAVPerceptualWeightsRead()",
                             argv[0]);
          QccErrorExit();
        }
      if (PerceptualWeights.num_subbands != NumSubbands)
        {
          QccErrorAddMessage("%s: Number of subbands in %s does not match specified number of levels (%d)",
                             argv[0],
                             PerceptualWeights.filename,
                             NumLevels);
        }
    }

  if (StepSize <= 0)
    {
      QccErrorAddMessage("%s: stepsize must be > 0",
                         argv[0]);
      QccErrorExit();
    }

  if (DeadZone < 0)
    {
      Quantizer.type = QCCSQSCALARQUANTIZER_UNIFORM;
      Quantizer.stepsize = StepSize;
    }
  else
    {
      Quantizer.type = QCCSQSCALARQUANTIZER_DEADZONE;
      Quantizer.stepsize = StepSize;
      Quantizer.deadzone = DeadZone;
    }

  if (QccIMGImageRead(&InputImage))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageRead()",
                         argv[0]);
      QccErrorExit();
    }
  if (QccIMGImageColor(&InputImage))
    {
      QccErrorAddMessage("%s: Color images not supported",
                         argv[0]);
      QccErrorExit();
    }
  if (QccIMGImageGetSize(&InputImage,
                         &ImageNumRows, &ImageNumCols))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageGetSize()",
                         argv[0]);
      QccErrorExit();
    }
  if (ImageNumCols != ((int)(ImageNumCols >> NumLevels) << NumLevels))
    {
      QccErrorAddMessage("%s: Image size (%d x %d) is not a multiple of %d (as needed for a decomposition of %d levels)",
                         argv[0],
                         ImageNumCols, ImageNumRows,
                         (1 << NumLevels), NumLevels);
      QccErrorExit();
    }
  if (ImageNumRows != ((int)(ImageNumRows >> NumLevels) << NumLevels))
    {
      QccErrorAddMessage("%s: Image size (%d x %d) is not a multiple of %d (as needed for a decomposition of %d levels",
                         argv[0],
                         ImageNumCols, ImageNumRows,
                         (1 << NumLevels), NumLevels);
      QccErrorExit();
    }
  
  OutputBuffer.type = QCCBITBUFFER_OUTPUT;
  if (QccBitBufferStart(&OutputBuffer))
    {
      QccErrorAddMessage("%s: Error calling QccBitBufferStart()",
                         argv[0]);
      QccErrorExit();
    }

   
  if (QccWAVsrEncode(&(InputImage.Y),
                     &OutputBuffer,
                     &Quantizer,
                     NumLevels,
                     &Wavelet,
                     ((UsePerceptualWeights) ? &PerceptualWeights : NULL)))
    {
      QccErrorAddMessage("%s: Error calling QccWAVsrEncode()",
                         argv[0]);
      QccErrorExit();
    }
      
  Rate = 
    (double)OutputBuffer.bit_cnt / ImageNumRows / ImageNumCols;

  if (QccBitBufferEnd(&OutputBuffer))
    {
      QccErrorAddMessage("%s: Error calling QccBitBufferEnd()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (ValueOnly)
    printf("%f\n", Rate);
  else
    {
      printf("SR coding of %s:\n",
             InputImage.filename);
      printf("  Total rate: %f bpp\n", Rate);
    }

  QccWAVWaveletFree(&Wavelet);
  QccWAVPerceptualWeightsFree(&PerceptualWeights);
  QccIMGImageFree(&InputImage);
  QccSQScalarQuantizerFree(&Quantizer);

  QccExit;
}
